File: java/java/spytools/net/spy/util/ThreadPool.java
// Copyright (c) 2001  Dustin Sallings <dustin@spy.net>
//
// $Id: ThreadPool.java,v 1.4 2002/02/24 07:25:57 dustin Exp $

package net.spy.util;

import java.util.*;

/**
 * A thread pool for easy parallelism.
 */
public class ThreadPool extends Object {

	// The threads we're managing.
	private Vector threads=null;
	// The tasks for the threads to do.
	private Stack tasks=null;

	// This is what we monitor for things being checked out (otherwise we
	// can't tell the difference between adds and check outs).
	private Object monitor=null;

	// Private thread ID allocator for the inner class.
	private static int thread_ids=0;

	/**
	 * Get an instance of ThreadPool.
	 *
	 * @param name Name of the pool.
	 * @param n Number of threads.
	 * @param prio Priority of the child threads.
	 */
	public ThreadPool(String name, int n, int priority) {
		super();

		if(priority<Thread.MIN_PRIORITY || priority>Thread.MAX_PRIORITY) {
			throw new IllegalArgumentException(priority
				+ " is an invalid priority.");
		}

		tasks=new Stack();
		threads=new Vector();
		monitor=new Object();

		// Get the group for this threadpool.
		ThreadGroup tg=new ThreadGroup(name);
		tg.setDaemon(true);

		// Initialize all of the threads.
		for(int i=0; i<n; i++) {
			RunThread rt=new RunThread(tg, tasks, monitor);
			rt.setPriority(priority);
			threads.addElement(rt);
		}
	}

	/**
	 * Get an instance of ThreadPool with a normal priority.
	 *
	 * @param name Name of the pool.
	 * @param n Number of threads.
	 */
	public ThreadPool(String name, int n) {
		this(name, n, Thread.NORM_PRIORITY);
	}

	/**
	 * Add a task for one of the threads to execute.
	 */
	public void addTask(Runnable r) {
		synchronized(tasks) {
			tasks.addElement(r);
			tasks.notify();
		}
	}

	/**
	 * Find out how many tasks are in the queue.
	 */
	public int getTaskCount() {
		int rv=0;
		synchronized(tasks) {
			rv=tasks.size();
		}
		return(rv);
	}

	/**
	 * Find out how many threads are in the pool.
	 */
	public int getThreadCount() {
		return(threads.size());
	}

	/**
	 * Tell all the threads to shut down after they finish their current
	 * tasks.
	 */
	public void shutdown() {
		for(Enumeration e=threads.elements(); e.hasMoreElements(); ) {
			RunThread t=(RunThread)e.nextElement();
			t.shutdown();
		}
	}

	/**
	 * Wait until there are no more than <i>num</i> tasks.
	 */
	public void waitForTaskCount(int num) throws InterruptedException {
		while(getTaskCount() > num) {
			synchronized(monitor) {
				monitor.wait(5000);
			}
		}
	}

	// A test task that takes a random amount of time.
	private static Runnable getTestRunnable() {
		return(new Runnable() {
			public void run() {
				try {
					Random rand=new Random();
					long l=Math.abs(rand.nextLong()%15000);
					System.err.println("Sleeping for " + l);
					Thread.sleep(l);
					System.err.println("Done!");
				} catch(Exception e) {
					e.printStackTrace();
				}
			}
			});
	}

	/**
	 * Shuts down in case you didn't.
	 */
	protected void finalize() throws Throwable {
		System.err.println(
			"********** Shutting down abandoned thread pool **********");
		shutdown();
	}

	/**
	 * Testing and what not.
	 */
	public static void main(String args[]) throws Exception {
		ThreadPool tp=new ThreadPool("TestThreadPool", 15);

		// Toss a hunded tasks into the thing.
		for(int i=0; i<100; i++) {
			tp.addTask(getTestRunnable());
			Thread.currentThread().getThreadGroup().list();
		}

		// Add another 100 tasks, but only if the task count is below 50
		for(int i=0; i<100; i++) {
			tp.waitForTaskCount(50);
			System.out.println("Adding new task.");
			tp.addTask(getTestRunnable());
		}

		// Wait for all of the tasks to finish
		tp.waitForTaskCount(0);
		System.out.println("All tasks have been accepted, shutting down.");
		// I shut 'em down!
		tp.shutdown();

		System.out.println("Done.");
	}

	// //////////////////////////////////////////////////////////////////////
	// The threads that make up the pool.
	// //////////////////////////////////////////////////////////////////////

	private class RunThread extends Thread {
		private Object monitor=null;
		private Stack tasks=null;
		private boolean going=true;
		private int thread_id=0;

		private String runningMutex=null;
		private Object running=null;
		private long start=0;

		public RunThread(ThreadGroup tg, Stack tasks, Object monitor) {
			super(tg, "RunThread");

			runningMutex=new String("runningMutex");
			this.tasks=tasks;
			this.monitor=monitor;

			thread_id=thread_ids++;

			System.out.println("RunThread " + thread_id + " going online.");

			// Adjust the name to include the thread number
			setName("RunThread#" + thread_id);
			// Note:  This should not be a daemon thread.
			start();
		}

		public String toString() {
			StringBuffer sb=new StringBuffer();
			sb.append(super.toString());

			synchronized(runningMutex) {
				if(running==null) {
					sb.append(" - idle");
				} else {
					sb.append(" - running ");
					sb.append(running.getClass().getName());
					sb.append(" for ");
					sb.append(System.currentTimeMillis() - start);
					sb.append("ms");
				}
			}
			return(sb.toString());
		}

		// I shut 'em down!
		public void shutdown() {
			going=false;
		}

		private void run(Runnable r) {
			try {
				// Record the runnable
				running=r;
				start=System.currentTimeMillis();
				// Run the runnable.
				r.run();
			} catch(Throwable t) {
				t.printStackTrace();
			}
			synchronized(runningMutex) {
				running=null;
			}
		}

		public void run() {
			while(going) {
				try {
					Runnable r=(Runnable)tasks.pop();
					// Let the monitor know we got one.
					synchronized(monitor) {
						monitor.notify();
					}
					run(r);
				} catch(EmptyStackException e) {
					// If the stack is empty, wait for something to get added.
					synchronized(tasks) {
						try {
							// Wait up to ten seconds
							tasks.wait(10000);
						} catch(InterruptedException ie) {
							// That's OK, we'll try again.
						}
					}
				} // empty stack
			} // while
		} // ThreadPool$RunThread.run()
	} // ThreadPool$RunThread
} // ThreadPool

File: python/threads/threadpool.py
#!/usr/bin/env python
#
# Copyright (c) 2002  Dustin Sallings <dustin@spy.net>
# $Id: threadpool.py,v 1.3 2002/03/28 20:08:27 dustin Exp $

import threading
import exceptions
import traceback
import time

def dumpThreads():
    """Dump all threads to stdout."""
    print "Threads:"
    for t in threading.enumerate():
        print "  * " + str(t)

class Job:
    """Superclass for all jobs."""
    def run():
        """This method will be called when the job is ready to be executed."""
        raise exceptions.NotImplementedError

class JobQueue:
    """This is a basic thread-safe stack."""

    def __init__(self):
        """Get a new JobQueue."""
        self.queue=list()
        self.mutex=threading.Lock()
        self.condition=threading.Condition(self.mutex)

    def append(self, job):
        """Add a new job to the queue."""
        if not isinstance(job, Job):
            raise ValueError("Job must be of class Job, not " + str(type(job)))
        
        try:
            self.mutex.acquire()
            self.queue.append(job)
            self.condition.notify()
        finally:
            # This could raise an exception if the mutex isn't locked.  I'm
            # not sure what I'd do about that, since things are probably
            # broken by then, anyway.
            self.mutex.release()

    def pop(self):
        """Get the next job from the queue.

        If no jobs are available, an IndexError will be raised.
        """

        rv=None
        try:
            self.mutex.acquire()
            rv=self.queue.pop()
            self.condition.notify()
        finally:
            self.mutex.release()
        return rv

    def __len__(self):
        """Find out how many unclaimed jobs are in this queue."""
        rv=None
        try:
            self.mutex.acquire()
            rv=len(self.queue)
        finally:
            self.mutex.release()
        return rv

    def wait(self, timeout=None):
        """Wait for an object to be added to or removed from the queue."""
        self.condition.acquire()
        self.condition.wait(timeout)
        self.condition.release()
        # time.sleep(timeout)

class RunThread(threading.Thread):
    """This is a worker thread.

    Pay no attention to the thread behind the curtain.
    """

    thread_id=0

    def __init__(self, queue):
        """Get a RunThread"""

        # super.init
        threading.Thread.__init__(self)

        # A reference to the object that this is currently running
        self.running=None
        # Mark the start time of a job
        self.startTime=0

        # Our reference to the queue
        self.queue=queue
        # This reminds us to keep going
        self.going=1

        self.thread_id=RunThread.thread_id
        RunThread.thread_id+=1

        print "RunThread " + str(self.thread_id) + " going online."

        self.setName("RunThread#" + str(self.thread_id))
        
        self.start()

    def runJob(self, job):
        try:
            self.startTime=time.time()
            self.running=job
            job.run()
        except:
            traceback.print_exc()
        self.running=None

    def isProcessing(self):
        """Return true if this thread is currently processing something."""
        return(self.running != None)

    def __str__(self):
        # Get the regular string representation, minus the > on the end.
        rv=threading.Thread.__repr__(self)[0:-1]

        if self.isProcessing():
            rv+=" - running " + str(self.running.__class__) + " for "
            rv+="%.2fs" % (time.time() - self.startTime)
        else:
            rv+=" - idle"
        rv+=">"
        return rv

    def run(self):
        while self.going:
            try:
                j=self.queue.pop()
                self.runJob(j)
            except IndexError:
                # print "Got an index error on pop."
                # Wait up to ten seconds for a job to be added
                self.queue.wait(10)

    def shutdown(self):
        """Shutdown this thread."""
        self.going=None

class ThreadPool:
    """A pool of threads that are eagerly awaiting work."""

    def __init__(self, name="AThreadPool", num=5):
        """Get a new thread pool."""
        self.queue=JobQueue()
        self.threads=list()

        for i in range(num):
            t=RunThread(self.queue)
            self.threads.append(t)

    def waitForTaskCount(self, n=0):
        """Wait for the number of tasks in the queue to reach the given
        number."""
        while len(self.queue) > n:
            self.queue.wait(5)

    def addTask(self, job):
        """Add a job to the queue."""
        self.queue.append(job)

    def getTaskCount(self):
        """How many jobs are in the queue?"""
        return len(self.queue)

    def getThreadCount(self):
        """How many threads are in the pool?"""
        return len(self.threads)

    def shutdown(self):
        """Shutdown the pool."""
        for t in self.threads:
            t.shutdown()

    def __del__(self):
        """Make sure the pool is shut down."""
        self.shutdown()

class SampleTask(Job):
    """A sample task for test purposes."""
    def run(self):
        """Sleep a random amount of time up to 15 seconds."""
        import random
        r=random.Random()
        st=r.random() * 15
        print "Sleeping for " + str(st) + " seconds."
        time.sleep(st)
        print "Done."

def main():
    tp=ThreadPool("Test Pool", 15)
    dumpThreads()
    try:
        for i in range(100):
            dumpThreads()
            tp.addTask(SampleTask())
    
        dumpThreads()
        for i in range(100):
            tp.waitForTaskCount(50)
            print "Adding a new task."
            tp.addTask(SampleTask())

        tp.waitForTaskCount(0)
        print "All tasks have been accepted, shutting down."
        tp.shutdown()
        dumpThreads()

        print "Done."
    except:
        traceback.print_exc()
        tp.shutdown()

if __name__ == '__main__':
    main()
